primitives/chain-polkadot-bulletin/src/lib.rs

```diff
@@ -63,6 +72,94 @@ pub const MAX_UNREWARDED_RELAYERS_IN_CONFIRMATION_TX: MessageNonce = 1024;
 /// Maximal number of unconfirmed messages at inbound lane for Cumulus-based parachains.
 pub const MAX_UNCONFIRMED_MESSAGES_IN_CONFIRMATION_TX: MessageNonce = 4096;
 
+/// This signed extension is used to ensure that the chain transactions are signed by proper
+pub type ValidateSigned = GenericSignedExtensionSchema<(), ()>;
+
+/// Signed extension schema, used by Polkadot Bulletin.
+pub type SignedExtensionSchema = GenericSignedExtension<(
+	(
+		CheckNonZeroSender,
+		CheckSpecVersion,
+		CheckTxVersion,
+		CheckGenesis<Hash>,
+		CheckEra<Hash>,
+		CheckNonce<Nonce>,
+		CheckWeight,
+	),
+	ValidateSigned,
+)>;
+
+/// Signed extension, used by Polkadot Bulletin.
+#[derive(Encode, Decode, Debug, PartialEq, Eq, Clone, TypeInfo)]
+pub struct SignedExtension(SignedExtensionSchema);
+
+impl sp_runtime::traits::SignedExtension for SignedExtension {
```

Comment:
Maybe we could have another trait something like `SignedExtensionSchemaProvider {SignedExtensionSchema = ...}` and implement `sp_runtime::traits::SignedExtension` automatically for the types that implement `SignedExtensionSchemaProvider`.

Reply:
That's a good idea, however it won't allow us to have a blanket implementation (I guess that is what you mean by "implement sp_runtime::traits::SignedExtension automatically", right?) of `sp_runtime::traits::SignedExtension`, because `sp_runtime::traits::SignedExtension` lives in the external crate (Substrate). And if we'll create some adapter for that, we won't be able to add custom methods  which we need here (e.g. `from_params` and so on). I know - that looks ugly, but probably in the future we'll switch to something better (auto detecting well known signed extensions and explicitly asking about unknown on relayer start? something that we'll need to think of in #2491)

Reply:
Oh yes, you're right. I forgot about that.